File: testbed_php/test/associations/OptionalNToOptionalNTest.php
<?php

class OptionalNToOptionalNTest extends UnitTestCase
{


  public function test_CreateNew()
  {
    $s = new StudentAE(99);
    $this->assertEqual(0,$s->numberOfMentors());
    
    $m = new MentorAE("blah");
    $this->assertEqual(0,$m->numberOfStudents());
  }
  

  public function test_addStudent()
  {
    $m = new MentorAE("blah");
    $s = new StudentAE(99);
    $m->addStudent($s);
    
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($s, $m->getStudent(0));
  }

  

  public function test_addStudentToMultipleMentors()
  {
    $m = new MentorAE("blah");
    $m2 = new MentorAE("blah2");
    
    
    $s = new StudentAE(99);
    $s2 = new StudentAE(992);
    $m->addStudent($s);
    $m2->addStudent($s2);
    $m2->addStudent($s);
    $s2->addMentor($m);
    
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m2,$s->getMentor(1));
    $this->assertEqual($m2,$s2->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(1));

    
    $this->assertEqual($s, $m->getStudent(0));
    $this->assertEqual($s2, $m->getStudent(1));
    $this->assertEqual($s2, $m2->getStudent(0));
    $this->assertEqual($s, $m2->getStudent(1));
  }
  

  public function test_addToNewMentor()
  {
    $m = new MentorAE("blah");
    $m2 = new MentorAE("blah2");
    $s = new StudentAE(99);
    
    $m->addStudent($s);
    $m2->addStudent($s);
    
    $this->assertEqual(2,$s->numberOfMentors());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m2,$s->getMentor(1));

    $this->assertEqual(1, $m->numberOfStudents());
    $this->assertEqual(1, $m2->numberOfStudents());

    $this->assertEqual($s, $m->getStudent(0));
    $this->assertEqual($s, $m2->getStudent(0));
  }  
  

  public function test_doNotRemoveFromExistingMentor()
  {
    $m = new MentorAE("blah");
    $m2 = new MentorAE("blah2");

    $s1 = new StudentAE(99);
    $s2 = new StudentAE(98);
    
    $m->addStudent($s1);
    $m->addStudent($s2);

    $this->assertEqual(2, $m->numberOfStudents());
    $this->assertEqual(0, $m2->numberOfStudents());
    
    $m2->addStudent($s1);

    $this->assertEqual(2, $m->numberOfStudents());
    $this->assertEqual($s1, $m->getStudent(0));
    $this->assertEqual($s2, $m->getStudent(1));
    
    $this->assertEqual(1, $m2->numberOfStudents());
    $this->assertEqual($s1, $m2->getStudent(0));
    
    $this->assertEqual($m,$s1->getMentor(0));
    $this->assertEqual($m2,$s1->getMentor(1));
    
    $this->assertEqual($m,$s2->getMentor(0));
  }   
  

  public function test_removeStudent()
  {
    $m = new MentorAE("blah");
    $s = new StudentAE(99);
    
    $m->addStudent($s);
    $m->removeStudent($s);
    
    $this->assertEqual(0,$s->numberOfMentors());
    $this->assertEqual(0,$m->numberOfStudents());
  }  


  public function test_deleteAll()
  {
    $m = new MentorAE("blah");
    $m2 = new MentorAE("blah");
    
    $s = new StudentAE(99);
    $s2 = new StudentAE(98);
    $s3 = new StudentAE(97);
    
    $m->addStudent($s);
    $m->addStudent($s2);
    $m->addStudent($s3);

    $m2->addStudent($s2);
    
    $m->delete();

    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual(1,$m2->numberOfStudents());
    
    $this->assertEqual(0,$s->numberOfMentors());
    $this->assertEqual(1,$s2->numberOfMentors());
    $this->assertEqual(0,$s3->numberOfMentors());
  }  
  

  public function test_isMaxNumber()
  {
    $this->assertEqual(3,StudentAE::maximumNumberOfMentors());
    $this->assertEqual(2,MentorAE::maximumNumberOfStudents());
  }
  

  public function test_addStudentsMax()
  {
    $m = new MentorAE("blah");
    $this->assertEqual(true,$m->addStudent(new StudentAE(99)));
    $this->assertEqual(true,$m->addStudent(new StudentAE(98)));
    $this->assertEqual(false,$m->addStudent(new StudentAE(97)));
  }


  public function test_addMentorsMax()
  {
    $s = new StudentAE(98);
    $this->assertEqual(true,$s->addMentor(new MentorAE("blah")));
    $this->assertEqual(true,$s->addMentor(new MentorAE("blah")));
    $this->assertEqual(true,$s->addMentor(new MentorAE("blah")));
    $this->assertEqual(false,$s->addMentor(new MentorAE("blah")));
  }

}

?>

File: testbed_ruby/test/associations/optional_n_to_optional_n_test.rb
require 'test/unit'
require './src-gen-umple/student_a_e'
require './src-gen-umple/mentor_a_e'
require './src-gen-umple/program_a_e'

module CruiseTest
class OptionalNToOptionalNTest < Test::Unit::TestCase

  def test_CreateNew
    s = StudentAE.new(99)
    assert_equal(0,s.number_of_mentors)
    
    m = MentorAE.new("blah")
    assert_equal(0,m.number_of_students)
  end
  
  def test_add_student
    m = MentorAE.new("blah")
    s = StudentAE.new(99)
    m.add_student(s)
    
    assert_equal(m,s.get_mentor(0))
    assert_equal(s, m.get_student(0))
  end

  def test_add_studentToMultipleMentors
    m = MentorAE.new("blah")
    m2 = MentorAE.new("blah2")
    
    
    s = StudentAE.new(99)
    s2 = StudentAE.new(992)
    m.add_student(s)
    m2.add_student(s2)
    m2.add_student(s)
    s2.add_mentor(m)
    
    assert_equal(m,s.get_mentor(0))
    assert_equal(m2,s.get_mentor(1))
    assert_equal(m2,s2.get_mentor(0))
    assert_equal(m,s2.get_mentor(1))

    
    assert_equal(s, m.get_student(0))
    assert_equal(s2, m.get_student(1))
    assert_equal(s2, m2.get_student(0))
    assert_equal(s, m2.get_student(1))
  end
  
  def test_addToNewMentor
    m = MentorAE.new("blah")
    m2 = MentorAE.new("blah2")
    s = StudentAE.new(99)
    
    m.add_student(s)
    m2.add_student(s)
    
    assert_equal(2,s.number_of_mentors)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m2,s.get_mentor(1))

    assert_equal(1, m.number_of_students)
    assert_equal(1, m2.number_of_students)

    assert_equal(s, m.get_student(0))
    assert_equal(s, m2.get_student(0))
  end  
  
  def test_doNotRemoveFromExistingMentor
    m = MentorAE.new("blah")
    m2 = MentorAE.new("blah2")

    s1 = StudentAE.new(99)
    s2 = StudentAE.new(98)
    
    m.add_student(s1)
    m.add_student(s2)

    assert_equal(2, m.number_of_students)
    assert_equal(0, m2.number_of_students)
    
    m2.add_student(s1)

    assert_equal(2, m.number_of_students)
    assert_equal(s1, m.get_student(0))
    assert_equal(s2, m.get_student(1))
    
    assert_equal(1, m2.number_of_students)
    assert_equal(s1, m2.get_student(0))
    
    assert_equal(m,s1.get_mentor(0))
    assert_equal(m2,s1.get_mentor(1))
    
    assert_equal(m,s2.get_mentor(0))
  end   
  
  def test_remove_student
    m = MentorAE.new("blah")
    s = StudentAE.new(99)
    
    m.add_student(s)
    m.remove_student(s)
    
    assert_equal(0,s.number_of_mentors)
    assert_equal(0,m.number_of_students)
  end  

  def test_deleteAll
    m = MentorAE.new("blah")
    m2 = MentorAE.new("blah")
    
    s = StudentAE.new(99)
    s2 = StudentAE.new(98)
    s3 = StudentAE.new(97)
    
    m.add_student(s)
    m.add_student(s2)
    m.add_student(s3)

    m2.add_student(s2)
    
    m.delete

    assert_equal(0,m.number_of_students)
    assert_equal(1,m2.number_of_students)
    
    assert_equal(0,s.number_of_mentors)
    assert_equal(1,s2.number_of_mentors)
    assert_equal(0,s3.number_of_mentors)
  end  
  
  def test_isMaxNumber
    assert_equal(3,StudentAE::maximum_number_of_mentors)
    assert_equal(2,MentorAE::maximum_number_of_students)
  end
  
  def test_add_studentsMax
    m = MentorAE.new("blah")
    assert_equal(true,m.add_student(StudentAE.new(99)))
    assert_equal(true,m.add_student(StudentAE.new(98)))
    assert_equal(false,m.add_student(StudentAE.new(97)))
  end


  def test_add_mentorsMax
    s = StudentAE.new(98)
    assert_equal(true,s.add_mentor(MentorAE.new("blah")))
    assert_equal(true,s.add_mentor(MentorAE.new("blah")))
    assert_equal(true,s.add_mentor(MentorAE.new("blah")))
    assert_equal(false,s.add_mentor(MentorAE.new("blah")))
  end
end
end
